x/delayedack/keeper/rollapp_packet.go

```diff
@@ -114,35 +120,44 @@ func (k *Keeper) UpdateRollappPacketWithStatus(ctx sdk.Context, rollappPacket co
 	return rollappPacket, nil
 }
 
-// ListRollappPendingPackets retrieves a list of pending rollapp packets from the KVStore.
+// ListRollappPacketsByStatus retrieves a list of pending rollapp packets from the KVStore.
 // It builds a prefix using the rollappID and the pending status, and iterates over the range from lastProofHeight to proofHeight.
 // If the packet's proofHeight is less than or equal to the maxProofHeight, it is added to the list.
+// if maxProofHeight is 0, all packets are returned.
 // The function returns the list of pending packets.
-func (k Keeper) ListRollappPendingPackets(
+func (k Keeper) ListRollappPacketsByStatus(
 	ctx sdk.Context,
-	rollappId string,
+	status commontypes.Status,
 	maxProofHeight uint64,
 ) (list []commontypes.RollappPacket) {
-	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(commontypes.RollappPacketKeyPrefix))
+	logger := ctx.Logger()
+	store := ctx.KVStore(k.storeKey)
 
-	// Build the prefix which is composed of the rollappID and the status
+	// switch prefix based on status
 	var prefix []byte
-	prefix = append(prefix, []byte(rollappId)...)
-	prefix = append(prefix, []byte("/")...)
-
-	// Append the pending status to the prefix
-	pendingStatusBytes := []byte(fmt.Sprint(commontypes.Status_PENDING))
-	prefix = append(prefix, pendingStatusBytes...)
-	prefix = append(prefix, []byte("/")...)
+	switch status {
+	case commontypes.Status_PENDING:
+		prefix = commontypes.PendingRollappPacketKeyPrefix
+	case commontypes.Status_FINALIZED:
+		prefix = commontypes.FinalizedRollappPacketKeyPrefix
+	case commontypes.Status_REVERTED:
+		prefix = commontypes.RevertedRollappPacketKeyPrefix
+	default:
+		return nil
+	}
 
 	// Iterate over the range from lastProofHeight to proofHeight
 	iterator := sdk.KVStorePrefixIterator(store, prefix)
 	defer iterator.Close() // nolint: errcheck
 
 	for ; iterator.Valid(); iterator.Next() {
 		var val commontypes.RollappPacket
-		k.cdc.MustUnmarshal(iterator.Value(), &val)
-		if val.ProofHeight <= maxProofHeight {
+		err := k.cdc.Unmarshal(iterator.Value(), &val)
+		if err != nil {
+			logger.Error("Failed to unmarshal rollapp packet", "error", err)
+			continue
+		}
+		if maxProofHeight == 0 || val.ProofHeight <= maxProofHeight {
```

Comment:
why not storing the PENDING keys by proofheight as well? to avoid iterating over all the PENDING packets when finalizing

Reply:
yea actually I think there is even a bug here which is worse. There is an assumption that the packets are ordered by proof height (which I think is a left over from pre-keys change) and on the first violation of proofHeight the loop breaks. which is bad .

Reply:
don't u want to add the proofheight to the iterator?

Reply:
it's iterating by order of proofHeight so once proofHeight > MaxProofHeight I just break the loop. 
trying to introduce endKey for the iterator is not straight forward as the proofHeight is in the middle of the rollappPacketKey bytes so can't be used as startKey/endKey.